xla/service/float_normalization_test.cc

```diff
@@ -691,4 +691,42 @@ TEST_F(FloatNormalizationTest, ConvertBeforeTuple) {
       F32);
 }
 
+TEST_F(FloatNormalizationNoComputeSupportTest, NoNormalizationForClonedModule) {
+  auto module = CreateNewVerifiedModule();
+  HloComputation::Builder sum_builder("sum");
+  auto x = sum_builder.AddInstruction(HloInstruction::CreateParameter(
+      /*parameter_number=*/0, ShapeUtil::MakeShape(BF16, {}), "x"));
+  auto y = sum_builder.AddInstruction(HloInstruction::CreateParameter(
+      /*parameter_number=*/1, ShapeUtil::MakeShape(BF16, {}), "y"));
+  sum_builder.AddInstruction(HloInstruction::CreateBinary(
+      ShapeUtil::MakeShape(BF16, {}), HloOpcode::kAdd, x, y));
+  HloComputation* reduction =
+      module->AddEmbeddedComputation(sum_builder.Build());
+
+  auto builder = HloComputation::Builder(TestName());
+  Shape bf16_shape_a = ShapeUtil::MakeShape(BF16, {2, 4});
+
+  HloInstruction* a = builder.AddInstruction(
+      HloInstruction::CreateParameter(0, bf16_shape_a, "a"));
+
+  HloInstruction* crs = builder.AddInstruction(
+      HloInstruction::CreateAllReduce(bf16_shape_a, {a}, reduction,
+                                      /*replica_groups=*/{},
+                                      /*constrain_layout=*/false,
+                                      /*channel_id=*/std::nullopt,
+                                      /*use_global_device_ids=*/false));
+
+  auto computation = module->AddEntryComputation(builder.Build());
+
+  auto cloned_module = module->Clone();
```

Comment:
Why do you need to clone here?

Reply:
The original failure is when the model is cloned, the IsCollectiveCalledComputation is not carried over properly which results in incorrect type-promotion. I'm adding that as a case here to verify cloning has no impact on the type-promotion bahavior.

Reply:
That seems like a convoluted way to test this pass: can't we directly create the desired IR to test without cloning? Is it even valid IR to have it without the IsCollectiveCalledComputation carried?

What if he cloned method if fixed to carry over the IsCollectiveCalledComputation? Is this test still relevant?

Reply:
The plan is to remove IsCollectiveCalledComputation completely, so I don't recommend trying to fix that.
That flag along with other flags like (isCustomCall and etc) are not part of the IR. They are only present in the HloInstruction object and relying on setters to set them correctly. The plan is to remove these flags and provide abstraction helpers to deduce from HLO graph directly.

Reply:
> The plan is to remove IsCollectiveCalledComputation completely

For all practical purpose here: this is a similar effect as to fixing the cloneModule though. Basically it means that "CloneModule"  is not relevant to this test as far as I can tell.

Thought experiment: let's say we remove these flags, is this test becoming useless?

> That flag along with other flags like (isCustomCall and etc) are not part of the IR. They are only present in the HloInstruction object and relying on setters to set them correctly. 

Ok, I guess we're using "part of the IR" differently here, but let's not get hung up on terminology: you have the HLOInstruction objects at hand here, so you could still set them in the test directly if it matters right? But does it matter for this test?

Reply:
I see where you are coming at, in terms of exercising code path, this test doesn't really add additional coverage compared to other tests in this file. And yes, when we remove the flags, this test is useless.
In fact, all the failures we see now are caused by incorrect values of the flag. I can specifically set an incorrect flag in this test and verify the HLO is correct even with a wrong flag. But it'd need to removed when we remove the flag.

Reply:
Right, so the advantage of setting `is_collective_called_computation_` flag explicitly is that indeed we'll be able to clean this up when the flag is removed. Instead of having an obsolete test with a weird cloneModule that in the future won't make sense.

Reply:
Well it looks like there's no easy way to manually set the flag to false outside of the class. I have tried mocking HloComputation with GMOCK, but it wont work when using that computation. For now, I change the test behavior to create a computation for allreduce and then replace the computation with another one that has the flag set to false. It's not the best but more explicit than using module->clone.